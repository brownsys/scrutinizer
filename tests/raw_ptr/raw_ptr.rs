#![allow(dead_code)]

mod raw_mut_ptr {
    // Raw mut pointer dereference.
    #[doc = "impure"]
    pub unsafe fn raw_mut_ptr_deref(a: usize) {
        let mut x = 42;
        let raw = &mut x as *mut i32;
        *raw = 5;
    }

    // Raw mut pointer aliasing.
    #[doc = "impure"]
    pub unsafe fn raw_mut_ptr_mut_ref(a: usize) {
        let mut x = 42;
        let raw = &mut x as *mut i32;
        let mut_ref = &mut *raw;
    }
}

mod raw_mut_ptr_call {
    #[derive(Debug)]
    struct Foo {
        x: i32
    }

    impl Foo {
        fn amend(&mut self) {
            self.x = 42;
        }
    }
    // Raw mut pointer dereference into call.
    #[doc = "impure"]
    pub unsafe fn raw_mut_ptr_deref_into_call(a: usize) {
        let mut x = Foo { x: 0 };
        let raw = &mut x as *mut Foo;
        (*raw).amend();
    }
}

mod safe_raw_mut_ptr {
    // Safe raw mut pointer dereference examples.
    #[doc = "pure"]
    pub unsafe fn safe_raw_mut_ptr_deref<'a>(a: usize) -> &'a i32 {
        let mut x = 42;
        let raw = &mut x as *mut i32;
        let immutable = *raw;
        &*raw
    }
}

mod raw_const_ptr {
    // Raw const pointer dereference.
    #[doc = "pure"]
    pub unsafe fn raw_const_ptr_deref(a: usize) {
        let x = 42;
        let raw = &x as *const i32;
        let _points_at = *raw;
    }
}

fn main() {}
