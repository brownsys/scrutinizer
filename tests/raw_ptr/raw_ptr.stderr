error: static analysis was not able to verify the purity of the region
  --> raw_ptr.rs:6:5
   |
6  | /     pub unsafe fn raw_mut_ptr_deref(a: usize) {
7  | |         let mut x = 42;
8  | |         let raw = &mut x as *mut i32;
9  | |         *raw = 5;
10 | |     }
   | |_____^
   |
   = help: consider using sandbox or privacy region
   = note: `#[deny(scrutinizer_lint)]` on by default

error: static analysis was not able to verify the purity of the region
  --> raw_ptr.rs:14:5
   |
14 | /     pub unsafe fn raw_mut_ptr_mut_ref(a: usize) {
15 | |         let mut x = 42;
16 | |         let raw = &mut x as *mut i32;
17 | |         let mut_ref = &mut *raw;
18 | |     }
   | |_____^
   |
   = help: consider using sandbox or privacy region

error: static analysis was not able to verify the purity of the region
  --> raw_ptr.rs:34:5
   |
34 | /     pub unsafe fn raw_mut_ptr_deref_into_call(a: usize) {
35 | |         let mut x = Foo { x: 0 };
36 | |         let raw = &mut x as *mut Foo;
37 | |         (*raw).amend();
38 | |     }
   | |_____^
   |
   = help: consider using sandbox or privacy region
